Fix logical error in elseif-statements in connect functions

janhahne opened the pull request:
Issue #1242 is a symptom of this logical error in the elseif-state in the connect functions in `connection_manager.cpp`. Currently the case `not source->has_proxies() and not target->has_proxies()` is handled before the case `not target->has_proxies() and not target->local_receiver()` which has a subcase `if ( not source->has_proxies() )` which therefore is not reachable.

This results in wrong behavior in a case where `not source->has_proxies() and not target->has_proxies() and not target->local_receiver()` holds (as for example in the reproducer of #1242). In addition a source code comment indicates that this case is not implemented at all at the moment. Therefore this PR also adds  a `throw IllegalConnection` here, to prevent a silent error where the connect call is executed without creating any connection.

jakobj commented:
thanks for catching this @janhahne. could you also please add a test based on #1242 ?

janhahne commented:
I added a regressiontest. However this can only ensure that such a connection cannot be created. If we decide to implement such connections at some point the reproducer from #1242 would be a good candidate to ensure consistent results with different numbers of threads.

stinebuu commented on nestkernel/connection_manager.cpp:
```diff
@@ -547,28 +548,29 @@ nest::ConnectionManager::connect( const index sgid,
     if ( not source->has_proxies() )
     {
+      throw IllegalConnection( "We do not allow to connect a device to a global receiver at the moment" );
       return false;
```
Same as above with the `return` statement.

stinebuu commented on nestkernel/connection_manager.cpp:
```diff
@@ -464,6 +464,18 @@ nest::ConnectionManager::connect( const index sgid,
+    if ( not source->has_proxies() )
+    {
+      throw IllegalConnection( "We do not allow to connect a device to a global receiver at the moment" );
+      return;
```
Maybe remove the `return` as I don't think it makes sense to return after an exception has been thrown?

stinebuu commented on testsuite/regressiontests/issue-1242.sli:
```diff
@@ -0,0 +1,53 @@
+M_ERROR setverbosity
+
+% Check if such a connections results in an exception
+{
```
Connection should be singular (so no s)

janhahne commented:
@stinebuu Thanks for the review! I addressed your comments. Regarding your question: As this is explicitly caught as a "special" case (before calling the regular connect) I just assumed that this cannot be the solution here, but I will give it a try.
